source/components/AblyTask/AblyTask.brs

```diff
@@ -9,53 +12,78 @@ sub init()
     ATTACHED: 11,
     MESSAGE: 15
   }
-  m.headers = {
-    "Accept": "application/json"
-    "Content-Type": "application/json"
-  }
 
+  ' Highest supported log level. If set higher it will be lowered to match.
   m.MAXIMUM_LOG_LEVEL = 5
 end sub
 
+' Any code run from this functions will be run in the async task thread
 sub runTask()
+  ' Get all the public values in one pass to limit the amount of rendezvous
   topValues = m.top.getFields()
+
+  ' Process constructor values
   m.ENDPOINT = topValues.endpoint
-  m.CHANNEL = topValues.channel
-  m.KEY = getConnectionKey()
+  m.CHANNELS = topValues.channels
   m.logLevel = m.top.logLevel
   if m.logLevel > m.MAXIMUM_LOG_LEVEL then m.logLevel = m.MAXIMUM_LOG_LEVEL
 
-  logInfo(m.KEY)
+  ' Get the Authentication key
+  m.authenticationKey = getAuthenticationKey()
+
+  ' Make sure we got something back to use as am authentication key
+  logInfo("Authentication Key:", m.authenticationKey)
   m.connectionKey = ""
-  if NOT isNonEmptyString(m.KEY) then return
+  if NOT isNonEmptyString(m.authenticationKey) then return
 
+  ' Establish the initial connection
   if connect() then
-    if attach() then
-      while stream()
-        sleep(20)
-      end while
-    end if
+    for each channel in m.CHANNELS
+      ' Send an attach requests for each channel
+      logInfo("Attaching:", channel, attach(channel))
+    end for
+
+    ' Start watching for events
+    while stream()
+      ' Add a small sleep to avoid CPU overheating
+      sleep(20)
+    end while
   end if
 end sub
 
 function connect() as Boolean
-  response = makeRequest(connectEndpoint(), Invalid, "GET", m.headers)
+  ' Make the initial connection request
+  response = makeRequest(connectEndpoint(), Invalid, "GET", {
+    "Accept": "application/json"
+    "Content-Type": "application/json"
+  })
+
+  ' Handle the response
   if response.code = 200 then
     if isNonEmptyArray(response.body) then
       firstEntry = response.body[0]
       if isNonEmptyAA(firstEntry) AND isNonEmptyAA(firstEntry.connectionDetails) then
+        ' Store the connection key and emit a connection event
+        logInfo("Connection:", connectionDetails)
         connectionDetails = firstEntry.connectionDetails
         m.connectionKey = connectionDetails.connectionKey
         m.top.connected = connectionDetails
         return true
       end if
     end if
   end if
+
+  ' There was an error, emit an error event
+  logError("Connection:", response.body)
+  m.top.error = response.body
   return false
 end function
 
-function attach() as Boolean
-  response = makeRequest(sendEndpoint(attachParameters()))
+function attach(channel as string) as Boolean
+  if channel = "" then return false
```

Comment:
Seems like a bit of an arbitrary check. As far as I can tell you can't _throw_ from a function in BrightScript, but I'm not sure checking for zero-length strings strengthens the interface any, and nor does it help debug issues at runtime either. For me this just adds noise and I would remove it.

Reply:
So you saying hit the endpoint and let it fail rather than short circuiting? That was the main reason for the check was to return early and not even bother hitting the services.

Reply:
Yes, for sure. Keep it as lightweight as possible.

In terms of whether it's safe to continue watching or not, that is really a separate concern. It's quite valid to install some kind of local listener for a particular channel but then to never actually successfully subscribe to it. In other words, the act of establishing a handler or listener within the local client is not tied to the remote async execution of registering a subscription with the server.

Reply:
a361c54